here-naksha-storage-http/src/main/java/com/here/naksha/storage/http/POpToQueryConverter.java

```diff
@@ -0,0 +1,175 @@
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.storage.http;
+
+import static com.here.naksha.lib.core.exceptions.UncheckedException.unchecked;
+import static com.here.naksha.lib.core.models.storage.POpType.*;
+
+import com.here.naksha.lib.core.models.storage.POp;
+import com.here.naksha.lib.core.models.storage.POpType;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.ArrayUtils;
+import org.jetbrains.annotations.NotNull;
+
+public class POpToQueryConverter {
+
+  public static final String EQ_OPERATOR = "=";
+  public static final String NOT_EQ_OPERATOR = "!=";
+  public static final String NULL = ".null";
+  public static final String AND_DELIMITER = "&";
+
+  public static final String OR_DELIMITER = ",";
+  public static final String PATH_SEGMENT_DELIMITER = ".";
+
+  private static final Map<POpType, String> SIMPLE_LEAF_OPERATORS = Map.of(
+      EQ, EQ_OPERATOR,
+      GT, "=gt=",
+      GTE, "=gte=",
+      LT, "=lt=",
+      LTE, "=lte=",
+      CONTAINS, "=cs=");
+
+  private POpToQueryConverter() {}
+
+  static String p0pToQuery(POp pOp) {
+    if (pOp.op() == AND) return and(pOp);
+    else return pOpToMultiValueComparison(pOp).resolve();
+  }
+
+  private static String and(POp pOp) {
+    assertHasAtLeastOneChildren(pOp);
+    return pOp.children().stream().map(POpToQueryConverter::p0pToQuery).collect(Collectors.joining(AND_DELIMITER));
+  }
+
+  private static MultiValueComparison pOpToMultiValueComparison(POp pOp) {
+    if (pOp.op() == AND) throw unsupportedOperation("AND can be only a top level operation");
+    if (pOp.op() == OR) return or(pOp);
+    if (pOp.op() == NOT) return not(pOp);
+    if (pOp.op() == EXISTS) return exists(pOp);
+    return simpleLeafOperator(pOp);
+  }
+
+  private static MultiValueComparison or(POp pOp) {
+    assertHasAtLeastOneChildren(pOp);
+
+    return pOp.children().stream()
+        .map(POpToQueryConverter::pOpToMultiValueComparison)
+        .reduce((l, r) -> {
+          if (!Objects.equals(l.operator, r.operator))
+            throw unsupportedOperation(
+                "Operators " + l.operator + " and " + r.operator + " combined in one OR");
+          if (!Objects.equals(l.path, r.path)) throw unsupportedOperation("Paths in OR are not equal");
+          return new MultiValueComparison(l.operator, l.path, ArrayUtils.addAll(l.values, r.values));
+        })
+        .orElseThrow(() -> new IllegalStateException("Should not reach here."));
+  }
+
+  private static MultiValueComparison not(POp pOp) {
+    assertHasNChildren(pOp, 1);
+    MultiValueComparison multiValueComparison =
+        pOpToMultiValueComparison(pOp.children().get(0));
+    String newOperator =
+        switch (multiValueComparison.operator) {
```

Comment:
There can be "not"+"exists" as well.

Reply:
operation "exists" is "!=.null" so its operator is "!="
in operation "not" we negate this operator so we get "=.null"
Therefore the only possible operators in operation "not" are "=" and "!=".

MultiValueComparison objects are not aware of POp operations concept like exists. They just holds the operator (the sign)

Reply:
Understood. Thanks!